### ISSUE TITLE
How should spack deal with packages that need to be installed as root?

### ISSUE THREAD

**Issue by krafczyk** [spack/spack#4432]
Some software packages such as `singularity` (currently working on adding it) need to be installed as root in order for them to work properly.

What should be the strategy spack takes for these packages?

Something which makes sense to me is to add a boolean member variable to the package class like `install_as_root`. This then be inspected on install to determine whether `sudo` should be invoked to install the package. Of course this scheme will only work with pre-defined build systems like `AutotoolsPackage` for `CMakePackage`, and ad-hoc build systems will be left out.

**Comment by adamjstewart** [comment]
Are you talking about [this singularity](http://singularity.lbl.gov/install-linux)? If so, I don't think you have to install it with root privileges. You only have to use `sudo` if you want to install it to `/usr/local`; Spack will install it to its own installation prefix. I've never seen a software package that couldn't be installed without root. If there were, how would developers test it?

**Comment by krafczyk** [comment]
I am referring to that singularity.

You can install it without sudo, and create images, however if you want to run images as your default user and not the root user, singularity needs to be installed with root.

**Comment by samfux84** [comment]
@adamjstewart We recently installed Singularity on one node in our cluster for having a test environment to play with it. In order to be usable, one needs to run the "make install" step with sudo, because the software contains a setuid binary.

From the documentation about the singularity config file, http://singularity.lbl.gov/docs-config:

```
ALLOW SETUID (boolean, default=’yes’)

This parameter toggles the global ability to execute the SETUID (SUID) portion of the code if it exists. As mentioned earlier, if the SUID features are disabled, various Singularity features will not function (e.g. mounting of the Singularity image file format).

You can however disable SUID support iff (if and only if) you do not need to use the default Singularity image file format and if your kernel supports user namespaces and you choose to use user namespaces.

note: as of the time of this writing, the user namespace is rather buggy
```

**Comment by alalazo** [comment]
> Some software packages such as singularity (currently working on adding it) need to be installed as root in order for them to work properly.

Another package of this kind is [likwid](https://github.com/RRZE-HPC/likwid)

**Comment by krafczyk** [comment]
@adamjstewart: what @samfux84 said.

**Comment by adamjstewart** [comment]
I see. This seems like it won't be very common, and isn't particularly specific to a single build system. I haven't tried it, but does the following code work?
```python
def install(self, spec, prefix):
    # Must be installed as root to work properly
    sudo = which('sudo')
    sudo('make', '-j{0}'.format(make_jobs), 'install')
```

**Comment by krafczyk** [comment]
So I tried something like that, but weirdly my password isn't being accepted. It's almost like stdin isn't being piped to the command or something except for carriage return.

I double checked that I can run sudo not through spack and it worked just fine.

**Comment by adamjstewart** [comment]
Yes, that's right, now I remember. Somewhere we are filtering all stdin and looking for `Ctrl+C` or `v`. I ran into this same problem when I was adding support for licensed files.

Perhaps the package could have a snippet that looks at the user name and raises an exception if it is not being installed as root? Then the user would run `sudo spack install <name>`.

**Comment by krafczyk** [comment]
so the primary concern I have with running `sudo spack install <name>`, is that wouldn't the database and other files that spack normally uses be tainted with root uid/gid and flags?

**Comment by adamjstewart** [comment]
Good point. We need some way to run `sudo make install` without running `sudo spack install`. This might be a lot more work than you imagine. I have no idea where to even get started on this.

**Comment by krafczyk** [comment]
Well that's depressing.. I mean I don't know what kind of filtering you're doing on stdin, but couldn't that stream be passed into the command if spack doesn't detect a `Ctrl+C` or `v`? Can you point me to the files which do the filtering? I'd like to have a look.

**Comment by adamjstewart** [comment]
I don't even know which files do the filtering. @alalazo @tgamblin?

**Comment by scheibelp** [comment]
FWIW: as part of RPM support, https://github.com/LLNL/spack/pull/631 supports redirecting a package install (e.g. DESTDIR). I don't know if it's 100% suitable as is: for example, it expects to redirect only the root of the DAG and doesn't handle relocating the package to its final location, so it wouldn't be useful if you want to automatically install a dependency (and then proceed with building dependents in the DAG), but it may be possible to adapt.

**Comment by tgamblin** [comment]
@krafczyk: what do you want the uid/gid of the files in the install dir to look like when the install completes?  Does the install directory need to be owned by root?  Is that the expectation?

As mentioned, I might need to rethink the input filtering for stuff like this.  Currently we don't make a huge effort to connect input to the forked build process.  Also, the mechanisms discussed so far in here assume `sudo` or something like it... is that something we can always assume?  I am not sure what the accepted way to do this on non-linux platforms is.

**Comment by tgamblin** [comment]
@scheibelp: I don't think `DESTDIR` helps here, as it won't fix the `setuid` binary issue.

**Comment by svenevs** [comment]
Depending on when / where the input filtering actually comes in, could you cheat it by forcing the password to be entered prior to all of this?

`spack install --sudo pkg`

During argument parsing, detect `--sudo`, pause and ask for password.  Then the rest of spack is unleashed?  That only helps if there is a very specific point in time that the input filtering gets enabled.

- If you fork at this point though, you get the same database problems as described above
- If you cache the password, that's a major security violation.

Presumably, if possible, the `package.py` would then have a way of flagging that `--sudo` is needed.

Related: If root-enabled installs are made possible, it might be a good idea to incorporate a mechanism for tracking exactly what files got put outside of `spack`.

Example: package X would need to install `/etc/profile.d/super_startup.sh`.

`requires("--sudo", ["/etc/profile.d/super_startup.sh"]) # or whatever`

Then be verbosely transparent: `spack uninstall --sudo pkg` enumerates exactly what files would be removed that are outside of spack?

>  Also, the mechanisms discussed so far in here assume sudo or something like it... is that something we can always assume? I am not sure what the accepted way to do this on non-linux platforms is.

Only Windows would be a problem.  But the user would have **already** needed to launch their command prompt as an administrator.  It would be way too difficult to use [`runas`](https://technet.microsoft.com/en-us/library/cc771525(v=ws.10).aspx), because there's no straightforward way to determine _who_ to `runas`.

On the other hand, a raw `sudo` may be inappropriate.  You probably want `sudo -i` to keep the input filtering available, but that has implications (performs full login).

**Comment by vsoch** [comment]
ping on a 5 year old issue! Just hit this - trying to install singularity and basic container functionality doesn't work without setuid, and I don't see any obvious setuid packages. Any quick fixes?

**Comment by samfux84** [comment]
For me, singularity somehow goes into a different category of software than other packages that can be installed with SPACK, as the installation contains setuid binaries. In our cluster, singularity is handled by the people that deploy the operating system on the compute nodes and not by me installing software for the end users with SPACK.

I would not recommend to maintain singularity the same way as software that can be installed without elevated privileges (root), but that is just my personal opinion.

**Comment by haampie** [comment]
- Using Spack as root is very dangerous.
- Making Spack ask for temporary root access sounds very dangerous: for the user it's unclear what the scope really is, it could hit bugs of the form `rm -rf $EMPTY_VARIABLE/bin`, ...

Maybe we could print some commands the user can run instead, so the scope of `sudo` is very limited and clear:

```
==> package: installed successfully
==> package: warning: executables in this package need to be root-owned and require setuid permissions. For security reasons, Spack does not grant those privileges. They should be set manually:
    sudo chown root:root "<prefix>/bin/suid-binary"
    sudo chmod u+s "<prefix>/bin/suid-binary"
```

In the end it's just a handful of packages that need this, and over time it should be fewer and fewer. For example `libfuse` has linux namespaces support, so it does not necessarily need a suid `fusermount`. Similarly `bubblewrap` used to be suid, but by now they're rootless. Docker has a rootless version too...

Edit: actually singularity already dumps a warning:

```
==> [2022-08-15-10:43:05.675367] Warning: 
            For full functionality, you'll need to chown and chmod some files
            after installing the package.  This has security implications.
            For details, see:
            https://sylabs.io/guides/2.6/admin-guide/security.html
            https://sylabs.io/guides/3.2/admin-guide/admin_quickstart.html#singularity-security

            We've installed a script that will make the necessary changes;
            read through it and then execute it as root (e.g. via sudo).

            The script is named:

            /home/harmen/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-11.1.0/singularity-3.8.5-3dwpbs37bva7dpyr5fj2bfaqcxxbihnq/bin/spack_perms_fix.sh
```
